### PR TITLE
Zarr-Python 3 compatibility 

### PR DESCRIPTION
**Pull request by jhamman**
- [x] Closes https://github.com/zarr-developers/zarr-python/issues/1953
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by jhamman** [comment]
We have a few more things to fix in Zarr before merging this. I'll circle back once this is ready for a review.

File: dask/array/core.py [hunk 1]
```diff
@@ -3632,12 +3633,15 @@ def from_zarr(
         if isinstance(url, os.PathLike):
             url = os.fspath(url)
         if storage_options:
-            store = zarr.storage.FSStore(url, **storage_options)
+            if Version(zarr.__version__) < Version("3.0.0.a0"):
+                store = zarr.storage.FSStore(url, **storage_options)
+            else:
+                zarr.store.RemoteStore(url, **storage_options)
```
Comment (jrbourbeau) [inline dask/array/core.py:3639]: Just a guess

```suggestion
                store = zarr.store.RemoteStore(url, **storage_options)
```

**Review by jrbourbeau — COMMENTED** [review]
Thanks @jhamman. Exciting stuff. Let me know if you need anything while working through the compatibility fixes

File: continuous_integration/scripts/install.sh [hunk 2]
```diff
@@ -27,7 +27,7 @@ if [[ ${UPSTREAM_DEV} ]]; then
         git+https://github.com/dask/distributed \
         git+https://github.com/dask/dask-expr \
         git+https://github.com/dask/fastparquet \
-        git+https://github.com/zarr-developers/zarr-python.git@main
+        git+https://github.com/zarr-developers/zarr-python.git@fix/dask-compat
```
Comment (jrbourbeau) [inline continuous_integration/scripts/install.sh:30]: I've added the `upstream` label to this PR which will now run a CI job with this branch from `zarr` when you push future commits here

Comment (jrbourbeau) [reply continuous_integration/scripts/install.sh:30]: You're probably already aware, but just posing the current crop of CI failures here for convenience 

```
FAILED dask/array/tests/test_array_core.py::test_zarr_existing_array - TypeError: object of type 'generator' has no len()
FAILED dask/array/tests/test_array_core.py::test_zarr_pass_store - ValueError: mode mismatch (mode=a != store.mode=AccessMode(str='w', readonly=False, overwrite=True, create=True, update=False))
FAILED dask/array/tests/test_array_core.py::test_zarr_group - ValueError: store mode does not support writing
FAILED dask/array/tests/test_array_core.py::test_zarr_regions - TypeError: object of type 'generator' has no len()
FAILED dask/tests/test_order.py::test_array_store_final_order - AttributeError: module 'zarr' has no attribute 'DirectoryStore'
FAILED dask/tests/test_sizeof.py::test_xarray_not_in_memory - TypeError: open_group() takes 0 positional arguments but 1 positional argument (and 4 keyword-only arguments) were given
FAILED dask/tests/test_distributed.py::test_zarr_distributed_with_explicit_memory_store - FileNotFoundError: memory://140106500045760
FAILED dask/tests/test_distributed.py::test_zarr_in_memory_distributed_err - TypeError: object of type 'generator' has no len()
ERROR dask/tests/test_distributed.py::test_zarr_distributed_roundtrip - Failed: 3 thread(s) were leaked from test

------ Call stack of leaked thread 1/3: <Thread(zarrIO, started daemon 140107162052160)> ------
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1032, in _bootstrap
	self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
	self.run()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1012, in run
	self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
	self._run_once()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/asyncio/base_events.py", line 1948, in _run_once
	event_list = self._selector.select(timeout)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/selectors.py", line 468, in select
	fd_event_list = self._selector.poll(timeout, max_ev)

------ Call stack of leaked thread 2/3: <Thread(asyncio_0, started daemon 140107587769920)> ------
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1032, in _bootstrap
	self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
	self.run()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1012, in run
	self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/concurrent/futures/thread.py", line 89, in _worker
	work_item = work_queue.get(block=True)

------ Call stack of leaked thread 3/3: <Thread(asyncio_1, started daemon 140107178833472)> ------
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1032, in _bootstrap
	self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
	self.run()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1012, in run
	self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/concurrent/futures/thread.py", line 89, in _worker
	work_item = work_queue.get(block=True)
ERROR dask/tests/test_distributed.py::test_zarr_distributed_with_explicit_directory_store - Failed: 2 thread(s) were leaked from test

------ Call stack of leaked thread 1/2: <Thread(asyncio_2, started daemon 140107145270848)> ------
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1032, in _bootstrap
	self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
	self.run()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1012, in run
	self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/concurrent/futures/thread.py", line 89, in _worker
	work_item = work_queue.get(block=True)

------ Call stack of leaked thread 2/2: <Thread(asyncio_3, started daemon 140106375620160)> ------
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1032, in _bootstrap
	self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1075, in _bootstrap_inner
	self.run()
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/threading.py", line 1012, in run
	self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/test-environment/lib/python3.12/concurrent/futures/thread.py", line 89, in _worker
	work_item = work_queue.get(block=True)
```

**Review by jrbourbeau — COMMENTED** [review]
Going to bump CI here as it looks like there's been some relevant changes over in https://github.com/zarr-developers/zarr-python/pull/2186 and I'm curious to see what tests looks like here now

**Comment by jhamman** [comment]
@jrbourbeau -- this is just about ready to go. Any idea what is going on with the upstream test build?

**Comment by jrbourbeau** [comment]
Hmm looking now...

**Review by jrbourbeau — COMMENTED** [review]
Okay, looks like it might be a `mamba=2`-related issue. Let's see if https://github.com/dask/dask/pull/11409 fixes things.

**Comment by jrbourbeau** [comment]
Alright, @jhamman merging `main` should fix CI

File: continuous_integration/scripts/install.sh (hunk 2 above)
Comment (jhamman) [inline continuous_integration/scripts/install.sh:30]:
```suggestion
        git+https://github.com/zarr-developers/zarr-python.git@v3
```

**Comment by jhamman** [comment]
So close here. We're down to just one error:

```
ERROR dask/tests/test_distributed.py::test_zarr_distributed_roundtrip - Failed: 4 thread(s) were leaked from test
```

@jrbourbeau - do you have thoughts on how to debug this? This is a part of the Dask test suite I don't quite understand.

**Comment by jhamman** [comment]
@martindurant - also pinging you here because I assume you have sorted something similar in the past w/ fsspec. Looking at the logs, I see 3-4 threads:

- `zarrIO` - our global IO thread where we run our IO loop
- `asyncio_{1,2,...}` - unclear where these are coming from.

**Comment by martindurant** [comment]
zarr.core.common.to_thread creates a hidden ThreadPoolExecutor and stuff gets assigned to run there. If you wanted to maintain this pool, you would have to create and close it in order not to be caught by dask's thread checker.

It is used throughout LocalStore, by the way, which makes no sense.

It is also used for offloading en/decoding, which makes more sense, as these can involve considerable CPU. However, dask already has a pool running, so I would tentatively suggest that dask should take care of parallelism, and zarr should have a mode in which it _never_ makes new threads.

There is even an argument that the ioloop should not be created in its thread since a (distributed) dask worker already has one in the main thread, but fsspec does this too. On a worker, you may well then have three asyncio loops running on different threads.

**Comment by mrocklin** [comment]
Heads-up, @jrbourbeau has been reassigned to baby-duty for the next month.  He's unlikely to be responsive here.  Your best bet may be @fjetter and colleagues.

**Comment by jhamman** [comment]
> zarr.core.common.to_thread creates a hidden ThreadPoolExecutor and stuff gets assigned to run there. If you wanted to maintain this pool, you would have to create and close it in order not to be caught by dask's thread checker.

Makes sense. I'm fairly confident that none of the threads here are from the `ThreadPoolExecutor` but I agree that we should probably be looking to manage this ourselves. 

> It is used throughout LocalStore, by the way, which makes no sense.

+1, I agree with this

> There is even an argument that the ioloop should not be created in its thread since a (distributed) dask worker already has one in the main thread, but fsspec does this too. On a worker, you may well then have three asyncio loops running on different threads.

I'm curious there are ways to detect this and utilize an existing thread (or loop) in a general way.  

Also, it just occurred to me that this test may also raise an error with fsspec but it is probably not exercised through the distributed test suite today. I think I can rig that up pretty easily...

**Comment by fjetter** [comment]
A _leaking_ thread is not a huge deal. For instance, we know that the offloading thread (I think Martin mentioned this earlier) is not cleaned up. We're initializing this on import time, see https://github.com/dask/distributed/blob/36020d6abe4506af08bae2807d78477b7916c8aa/distributed/utils_test.py#L112 such that the thread is already up and running before the test runs. So, if there is a way to initialize that hidden threadpool once before the tests run, that would be the cleanest solution

**Comment by jhamman** [comment]
@fjetter and co -- this is ready for a real review. The leaked thread issue was mostly fixed upstream, now Zarr includes a `cleanup` utility for its internal threads.

**Review by phofl — APPROVED** [review]
lgtm

**Review by jrbourbeau — COMMENTED** [review]
Nice, glad to see this got over the finish line. Thanks all

File: dask/tests/test_sizeof.py [hunk 3]
```diff
@@ -276,12 +277,19 @@ def test_xarray():
...
 def test_xarray_not_in_memory():
     xr = pytest.importorskip("xarray")
+    zarr = pytest.importorskip("zarr")
     np = pytest.importorskip("numpy")
     pytest.importorskip("zarr")
 
     ind = np.arange(-66, 67, 1).astype(float)
     arr = np.random.random((len(ind),))
 
+    # TODO: remove this conditional after consolidated metadata lands in v3
```
Comment (QuLogic) [inline dask/tests/test_sizeof.py:287]: Is this referring to landing in dask or in Zarr? Did it land? There's no issue or PR reference to check.

File: dask/tests/test_sizeof.py [hunk 4]
```diff
@@ -276,12 +277,19 @@ def test_xarray():
 
 def test_xarray_not_in_memory():
     xr = pytest.importorskip("xarray")
+    zarr = pytest.importorskip("zarr")
     np = pytest.importorskip("numpy")
     pytest.importorskip("zarr")
```
Comment (QuLogic) [inline dask/tests/test_sizeof.py:282]: PS, the second `zarr` import can be removed.

**Comment by biprateep** [comment]
Hi,
With the added support for Zarr-Python 3, will the [chunk-sharding feature](https://zarr.readthedocs.io/en/latest/user-guide/arrays.html#sharding)  be available for `arr.to_zarr()`? Thanks!

**Pull request merged**